[bazel] added Bazel getting started to docs (#11435)
lowRISC/opentitan: doc/ug/getting_started_build_sw.md

```console
sudo apt install apt-transport-https curl gnupg
curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor > bazel.gpg
sudo mv bazel.gpg /etc/apt/trusted.gpg.d/
echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" |
sudo tee /etc/apt/sources.list.d/bazel.list
sudo apt update && sudo apt install bazel-4.2.0
```

or by following [instructions for your system](https://bazel.build/install).

### Building software with Bazel

You can then build and run all the tests all the Bazel rules for OpenTitan within the workspace with the command:

```console
bazel test //...
```

This is likely to include unhealthy or slow tests and build targets so you're likely to want to run more specific builds and tests.
This invocation instructs Bazel to build everything it needs to run all the tests in the workspace.

Specific targets can be built with:

```console
bazel build //<path to build file>:<name of target>
```

### Structure of our Bazel workspace

The rules for Bazel are described in a language called starlark, which looks a lot like Python.

The OpenTitan directory is defined as a Bazel workspace by the ```//WORKSPACE``` file.

> ``BUILD``` files provide the information Bazel needs to build the targets in a directory.

Author: Thanks, I'll change this to single quotes for inline code/paths

### Linting Starlark

The Linter for Bazel files is called "buildifier" and can be built and run with Bazel by running:

```console
bazel run //buildifier_fix
```

Which will run the linter on all of the WORKSPACE,BUILD and .bzl files in the workspace.
It's fairly strict so the best way to keep git logs and blame reports clean is to run it before committing, and ask the same during code review.

### Test Execution

Bazel will execute tests when invoked with ```bazel test <label-expression>```.
The test rules have various tags applied to them to allow for filtering of tests during dependency analysis.
Our tests are typically tagged with verilator, cw310, or manual.
The ```--test_tag_filters``` switch can be used to filter out tests that you either cannot or do not want to execute.

For example, if you do not have a CW310 FPGA board, you cannot execute the tests meant to execute on that board.
You can instruct Bazel to not execute those tests:

```console
bazel test --test_tag_filters=-cw310 //...
```

> Rather than manually adding that command line switch to every invocation of bazel, you can add it to your $HOME/.bazelrc file.

Reviewer: ```suggestion
Rather than manually adding that command line switch to every invocation of bazel, you can add it to your `$HOME/.bazelrc` file.
```

> $HOME/.bazelrc:

Reviewer: I think this needs to be wrapped in a code block.

Author: Yeah, I was wary of mixing the path specification with the code, but you're right and I've already described the path inline so I'll use that

test --test_tag_filters=-cw310

### Excluding Verilator

Verilator is slow to build and Verilator tests are slow to run.
Many directories have Verilator targets that will cause them to build Verilator when you build all targets with the wildcard `...`.
Tagging allows us to exclude Verilator tests, but doesn't exclude building Verilator for targets that depend on it.
To exclude all targets that depend on Verilator you should use a target pattern that can be generated with Bazel queries, and sed and then passed with xargs:

```console
bazel query "rdeps(//..., //hw:verilator)" \
| sed -e 's/^/-/' \
| xargs bazel test -- //...
```

The Bazel query lists all the dependencies in the workspace that depend on Verilator.
These are inverted for the target pattern by prepending a dash and then the test is invoked by xargs on a pattern which starts with the whole workspace and then excludes each of the targets that depend on Verilator.

### Caching on disk

By default Bazel caches in memory and will evict Verilator binaries often.

> If you're regularly running tests that depend on Verilator and have a few GB of disk space available use the disk_cache arg to specify a cache:

Reviewer: ```suggestion
If you're regularly running tests that depend on Verilator and have a few GB of disk space available use the `--disk_cache` argument to specify a cache:
```

If you do not specify your own toolchain configuration file (using `./meson_init.sh -t`), and `meson_init.sh` cannot find the default configuration in your toolchain, the legacy `toolchain.txt` from the main OpenTitan repository will be used.
If `TOOLCHAIN_PATH` is set, this will be used to update any paths within the legacy configuration.

## Building Software with Bazel

The Mask ROM and many other software targets are built with
(Bazel)[https://bazel.build/]

TLDR:

> after installing Bazel you can run

Reviewer: nit: colon at end of line (other lines should similarly end in either a colon or a full stop).

```console
bazel test //... --test_tag_filters=-cw310 --disk_cache=~/bazel_cache
```

### Installing Bazel

> If you haven't yet installed Bazel, add it to apt and instal it on ubuntu systems  with the following

Reviewer: ```suggestion
If you haven't yet installed Bazel, add it to apt and install it on ubuntu systems with the following commands:
```

> commands:

Reviewer: Please use one sentence per line (i.e. no new lines mid sentence) to be consistent with the other markdown. This applies throughout.

```console
sudo apt install apt-transport-https curl gnupg
curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor > bazel.gpg
sudo mv bazel.gpg /etc/apt/trusted.gpg.d/

> echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" |

Reviewer: Might be worth mentioning somewhere that the jdk1.8 thing is a legacy thing and doesn't actually mean jdk1.8 is being used, just to stop people going down that rabbit hole.

Author: I've linked to the discussion upstream in the Bazel documentation. I hope that's sufficient.

sudo tee /etc/apt/sources.list.d/bazel.list
sudo apt update && sudo apt install bazel-4.2.0
```

or by following [instructions for your system](https://bazel.build/install)

> though it may not be supported by OpenTitan.

Reviewer: ```suggestion
```
I don't think we need this caveat.

### Building software with Bazel

> You can then build and run all the tests all the Bazel rules for OpenTitan within the workspace with the command

Reviewer: ```suggestion
You can then build and run all the tests all the Bazel rules for OpenTitan within the workspace with the command:
```

```console
bazel test //...
```

> This is likely to include unhealthy tests and targets and slow tests and targets

Reviewer: ```suggestion
This is likely to include unhealthy or slow tests and build targets
```

> so you're likely to want to run more specific builds and tests, but this

Reviewer: ```suggestion
so you're likely to want to run more specific builds and tests, however this
```

Author: Sorry for the comma splice. Unwrapped sentences really shows off my run-on sentences.

invocation instructs bazel to build everything it needs to run all the tests in
the workspace.

> Specific targets can be built with

Reviewer: ```suggestion
Specific targets can be built with:
```

Author: All introductions to console segments will be punctuated with a colon

```console
bazel build //<path to build file>:<name of target>
```

### Structure of our Bazel workspace

The rules for Bazel are described in a language called starlark, which looks an
awful lot like Python.

> The OpenTitan directory is defined as a bazel workspace by the

Reviewer: ```suggestion
The OpenTitan directory is defined as a Bazel workspace by the
```
nit: should consistently use either Bazel or bazel, not a mix.

build the targets in the directories they reside in.

OpenTitan uses .bzl files
to specify rules to build targets for platforms that require specific attention
like generation of disassembly files and different compilers.

#### WORKSPACE file

The Workspace file controls external dependencies such that builds can be made
reproducible and hermetic.
Bazel loads specific commits of external git repositories and uses them to build
OpenTitan targets (like it does with bazel\_embedded) or to satisfy dependencies
(as it does with abseil). As we produce increasingly stable releases we should
fix a number of these references to specific commits and not the default branch
head.

#### BUILD files

Throughout the OpenTitan repo, BUILD files describe targets and dependencies in
the same directory and subdirectories that lack their own BUILD files. We don’t
currently have plans to automatically generate BUILD files, and to suit the
request we don’t include handwritten files in autogen directories, you’ll see
some big BUILD files that describe how to build and depend on auto generated
files in autogen directories.

### Linting Starlark

> The Linter for Bazel files is called “buildifier” and can be built and run with

Reviewer: ```suggestion
The Linter for Bazel files is called "buildifier" and can be built and run with
```

Author: I'll get rid of all the smart quotes I can find

Reviewer 2: There's a handy grep command in `ci/scripts/check-ascii.sh`. We allow non-ASCII in Markdown files at the moment, so running that script won't catch anything, but I think you probably just need something like `env LC_ALL=C xargs grep -d skip -P '[^\0-\x7f]' my-markdown-file.md`.

tests:

```console
bazel test --test_tag_filters=-cw310 //...
```

Rather than manually adding that command line switch to every invocation of
bazel, you can add it to your $HOME/.bazelrc file.

$HOME/.bazelrc:
test --test_tag_filters=-cw310

### Excluding Verilator

Verilator is slow to build and Verilator tests are slow to run. Many directories
have Verilator targets that will cause them to build Verilator when you build
all targets with the wildcard `...`. Tagging allows us to exclude verilator
tests, but doesn't exclude building verilator for targets that depend on it. To
exclude all targets that depend on verilator you should use a target pattern
that can be generated with Bazel queries, and sed and then passed with xargs:

```console
bazel query "rdeps(//..., //hw:verilator)" \
| sed -e 's/^/-/' \
| xargs bazel test -- //...
```

The bazel query lists all the dependencies in the workspace that depend on
verilator. These are inverted for the target pattern by prepending a dash and
then the test is invoked by xargs on a pattern which starts with the whole
workspace and then excludes each of the targets that depend on verilator.

### Caching on disk

> If you’re building verilator and you’re tired of waiting for it to build and

Reviewer: nit: I think this sentence could be shortened.

Author: Sorry for editorializing
